### ISSUE TITLE
Failed NCL diagnostic script does not exit but hangs

### ISSUE THREAD

**Issue by valeriupredoi** [ESMValGroup/ESMValCore#531]
**Describe the bug**
Clean exit after failed NCL diagnostic fails if in multiprocessing, resulting in hanging
First discovered via new integration test in #514 

**Steps to reproduce**
- install the latest version of `esmvalcore`
- set `max_parallel_tasks: ` > 1 in config-user
- run an NCL diagnostic script that fails for any particular reason

**Tracebacks**
```
2020-03-02 15:33:29,017 UTC [6285] INFO    Maximum memory used (estimate): 0.9 GB
2020-03-02 15:33:29,018 UTC [6285] INFO    Sampled every second. It may be inaccurate if short but high spikes in memory consumption occur.
2020-03-02 15:33:29,018 UTC [6285] ERROR   Program terminated abnormally, see stack trace below for more information
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/site-packages/esmvalcore/_task.py", line 694, in _run_task
    output_files = task.run()
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/site-packages/esmvalcore/_task.py", line 246, in run
    self.output_files = self._run(input_files)
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/site-packages/esmvalcore/_task.py", line 507, in _run
    self._control_ncl_execution(process, last_line + lines)
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/site-packages/esmvalcore/_task.py", line 411, in _control_ncl_execution
    raise DiagnosticError(msg)
esmvalcore._task.DiagnosticError: An error occurred during execution of NCL script /home/valeriu/esmvaltool_workshop_21Jan2019/plop.ncl, see the log in /home/valeriu/esmvaltool_workshop_21Jan2019/dryrun/recipe_dask_20200302_153310/run/validation_basic/ipython/log.txt
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/site-packages/esmvalcore/_main.py", line 220, in run
    conf = main(args)
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/site-packages/esmvalcore/_main.py", line 156, in main
    process_recipe(recipe_file=recipe, config_user=cfg)
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/site-packages/esmvalcore/_main.py", line 206, in process_recipe
    recipe.run()
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/site-packages/esmvalcore/_recipe.py", line 1320, in run
    run_tasks(self.tasks,
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/site-packages/esmvalcore/_task.py", line 615, in run_tasks
    _run_tasks_parallel(tasks, max_parallel_tasks)
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/site-packages/esmvalcore/_task.py", line 661, in _run_tasks_parallel
    _copy_results(task, running[task])
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/site-packages/esmvalcore/_task.py", line 682, in _copy_results
    task.output_files, updated_products = future.get()
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/pool.py", line 768, in get
    raise self._value
esmvalcore._task.DiagnosticError: An error occurred during execution of NCL script /home/valeriu/esmvaltool_workshop_21Jan2019/plop.ncl, see the log in /home/valeriu/esmvaltool_workshop_21Jan2019/dryrun/recipe_dask_20200302_153310/run/validation_basic/ipython/log.txt
2020-03-02 15:33:29,021 UTC [6285] INFO    If you suspect this is a bug or need help, please open an issue on https://github.com/ESMValGroup/ESMValTool/issues and attach the run/recipe_*.yml and run/main_log_debug.txt files from the output directory.

```
(it is in a hanging state) - upon Ctrl-C-ing:
```
^CProcess ForkPoolWorker-2:
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/util.py", line 277, in _run_finalizers
    finalizer()
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/util.py", line 201, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/pool.py", line 689, in _terminate_pool
Traceback (most recent call last):
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/pool.py", line 114, in worker
    task = get()
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/queues.py", line 355, in get
    with self._rlock:
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/synchronize.py", line 95, in __enter__
    return self._semlock.__enter__()
KeyboardInterrupt
    cls._help_stuff_finish(inqueue, task_handler, len(pool))
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/pool.py", line 674, in _help_stuff_finish
    inqueue._rlock.acquire()
KeyboardInterrupt
Process ForkPoolWorker-1:
Traceback (most recent call last):
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/pool.py", line 114, in worker
    task = get()
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/queues.py", line 356, in get
    res = self._reader.recv_bytes()
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/connection.py", line 216, in recv_bytes
    buf = self._recv_bytes(maxlength)
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/connection.py", line 414, in _recv_bytes
    buf = self._recv(4)
  File "/home/valeriu/anaconda3R/envs/esmvaltool_py/lib/python3.8/multiprocessing/connection.py", line 379, in _recv
    chunk = read(handle, remaining)
KeyboardInterrupt
```

**Comment by valeriupredoi** [comment]
The same problem these guys are seeing: https://bugs.python.org/issue38501 -> seems to be introduced by https://github.com/python/cpython/pull/11488

**Comment by valeriupredoi** [comment]
solved by #534
